crates/ruff_cli/src/commands/format.rs

```diff
@@ -35,41 +44,64 @@ pub(crate) fn format(cli: &FormatArguments, overrides: &Overrides) -> Result<Exi
         return Ok(ExitStatus::Success);
     }
 
-    let result = paths
+    let start = Instant::now();
+    let (results, errors): (Vec<_>, Vec<_>) = paths
         .into_par_iter()
         .map(|entry| {
             let entry = entry?;
             let path = entry.path();
-            if matches!(
-                SourceType::from(path),
-                SourceType::Python(PySourceType::Python | PySourceType::Stub)
-            ) {
+            if let SourceType::Python(source_type @ (PySourceType::Python | PySourceType::Stub)) =
+                SourceType::from(path)
+            {
                 let line_length = resolver.resolve(path, &pyproject_config).line_length;
-                let options = PyFormatOptions::from_extension(path)
+                let options = PyFormatOptions::from_source_type(source_type)
                     .with_line_width(LineWidth::from(NonZeroU16::from(line_length)));
-
                 format_path(path, options)
             } else {
-                Ok(())
+                Ok(FormatResult::Skipped)
             }
         })
-        .map(|result| {
-            result.map_err(|err| {
-                err.show_user();
-                err
-            })
-        })
-        .collect::<Result<(), _>>();
+        .partition_map(|result| match result {
+            Ok(diagnostic) => Left(diagnostic),
+            Err(err) => Right(err),
+        });
+    let duration = start.elapsed();
+    debug!("Formatted files in: {:?}", duration);
+
+    // Report on any errors.
+    if !errors.is_empty() {
+        warn!("Encountered {} errors while formatting:", errors.len());
+        for error in &errors {
+            error.show_user();
+        }
+    }
+
+    // Report on the formatting changes.
+    if log_level >= LogLevel::Default {
+        let mut writer: Box<dyn Write> = match &cli.output_file {
+            Some(path) => {
+                colored::control::set_override(false);
+                let file = File::create(path)?;
+                Box::new(BufWriter::new(file))
+            }
+            _ => Box::new(BufWriter::new(io::stdout())),
+        };
+        let summary = FormatResultSummary::from(results);
+        summary.show_user(&mut writer)?;
```

Comment:
Nit: Implement `Display` for summary instead so that you can do `write!(writer, "{summary}")`?